### PR TITLE
CORE-19679 Add simple subscription message pattern

### PR DESCRIPTION
**Pull request by owenstanford**
This change adds a new messaging pattern to the messaging library. The pattern is intended for use cases where the client wants a simple subscription only with auto offset committing.

### CODE REVIEW

File: libs/messaging/messaging-impl/src/main/kotlin/net/corda/messaging/subscription/EventSourceConsumer.kt [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    private val errorMsg =
+        "Failed to read and process records from topic=${topic}, group=${group}. Recreating consumer and Retrying."
+
+    private var cordaConsumer: CordaConsumer<K, V>? = null
+
+    /**
+     * Polls the event source consumer.
+     */
+    fun poll() {
+        if (cordaConsumer == null) {
```
Comment (yift-r3) [inline libs/messaging/messaging-impl/src/main/kotlin/net/corda/messaging/subscription/EventSourceConsumer.kt]: AtomicReference would probably be better than a var, but if we want to keep the var, we can do something like:
```
        val consumer = cordaConsumer.let {
            it ?: eventSourceCordaConsumerFactory.create().also {
                cordaConsumer = it
            }
        }
```
And remove the need for the `checkNotNull`

File: libs/messaging/messaging-impl/src/main/kotlin/net/corda/messaging/subscription/EventSourceConsumer.kt [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+        val consumer = checkNotNull(cordaConsumer)
+
+        try {
+            eventSourceRecordConsumer.poll(consumer)
+        } catch (ex: Exception) {
+            when (ex) {
+                // If we receive an intermittent exception we dispose the existing
+                // consumer and then let the poller call again to recreate a new one
+                is CordaMessageAPIIntermittentException -> {
+                    log.warn("$errorMsg ", ex)
```
Comment (yift-r3) [inline libs/messaging/messaging-impl/src/main/kotlin/net/corda/messaging/subscription/EventSourceConsumer.kt]: Why not `log.warn(errorMsg, ex)`?

File: libs/messaging/messaging-impl/src/main/kotlin/net/corda/messaging/subscription/EventSourceConsumer.kt [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+ */
+internal class EventSourceConsumer<K : Any, V : Any>(
+    group: String,
+    topic: String,
+    private val eventSourceCordaConsumerFactory: EventSourceCordaConsumerFactory<K, V>,
+    private val eventSourceRecordConsumer: EventSourceRecordConsumer<K, V>,
+    private val log: Logger
+) {
+
+    private val errorMsg =
```
Comment (yift-r3) [inline libs/messaging/messaging-impl/src/main/kotlin/net/corda/messaging/subscription/EventSourceConsumer.kt]: Is keeping the string helpful? If so, could it be lazy (i.e. create it only when we need it the first time)?

Anyway - `topic=${topic}, group=${group}` can be `topic=$topic, group=$group`

File: libs/messaging/messaging-impl/src/main/kotlin/net/corda/messaging/subscription/EventSourceConsumer.kt [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+
+        val consumer = checkNotNull(cordaConsumer)
+
+        try {
+            eventSourceRecordConsumer.poll(consumer)
+        } catch (ex: Exception) {
+            when (ex) {
+                // If we receive an intermittent exception we dispose the existing
+                // consumer and then let the poller call again to recreate a new one
+                is CordaMessageAPIIntermittentException -> {
```
Comment (yift-r3) [inline libs/messaging/messaging-impl/src/main/kotlin/net/corda/messaging/subscription/EventSourceConsumer.kt]: Why not `catch (e: CordaMessageAPIIntermittentException) {...` and then `catch (e: Exception) {...`?

File: libs/messaging/messaging-impl/src/main/kotlin/net/corda/messaging/subscription/ThreadLooper.kt [hunk 5]
```diff
@@ -166,6 +172,33 @@ class ThreadLooper(
...
+     * The continuously running poll loop will keep calling the loop function until the looper is marked as stopped
+     * The Looper will mark itself as up if it can complete the first poll successfully.
+     */
+    private fun runContinuousPollLoop() {
+        try {
+            var isFirstPoll = true
+            while (_isRunning) {
+                loopFunction()
+                if (isFirstPoll) {
+                    lifecycleCoordinator.updateStatus(LifecycleStatus.UP)
```
Comment (yift-r3) [inline libs/messaging/messaging-impl/src/main/kotlin/net/corda/messaging/subscription/ThreadLooper.kt:197]: Can the loop function set the state to error? If so, we need to check that the state is not an error before we set it up...

File: libs/messaging/messaging-impl/src/main/kotlin/net/corda/messaging/subscription/ThreadLooper.kt [hunk 6]
```diff
@@ -166,6 +172,33 @@ class ThreadLooper(
...
+     *
+     * The continuously running poll loop will keep calling the loop function until the looper is marked as stopped
+     * The Looper will mark itself as up if it can complete the first poll successfully.
+     */
+    private fun runContinuousPollLoop() {
+        try {
+            var isFirstPoll = true
+            while (_isRunning) {
+                loopFunction()
+                if (isFirstPoll) {
```
Comment (yift-r3) [inline libs/messaging/messaging-impl/src/main/kotlin/net/corda/messaging/subscription/ThreadLooper.kt:196]: Should we check isRunning as well?

File: libs/messaging/messaging-impl/src/main/kotlin/net/corda/messaging/subscription/factory/CordaSubscriptionFactory.kt [hunk 7]
```diff
@@ -166,6 +173,60 @@ class CordaSubscriptionFactory @Activate constructor(
...
+        val consumerFactory = EventSourceCordaConsumerFactory(
+            config.group,
+            config.clientId,
+            config.topic,
+            processor.keyClass,
+            processor.valueClass,
+            config.messageBusConfig,
+            partitionAssignmentListener,
+            cordaConsumerBuilder,
+            LoggerFactory.getLogger("${EventSourceCordaConsumerFactory<K, V>::javaClass.name}-${config.clientId}")
```
Comment (yift-r3) [inline libs/messaging/messaging-impl/src/main/kotlin/net/corda/messaging/subscription/factory/CordaSubscriptionFactory.kt:197]: We can create the loggers inside the class instead of passing it? Are we creating them here for unit testing?

File: libs/messaging/messaging-impl/src/main/kotlin/net/corda/messaging/subscription/factory/EventSourceCordaConsumerFactory.kt [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+import net.corda.messaging.subscription.consumer.listener.ForwardingRebalanceListener
+import net.corda.messaging.subscription.consumer.listener.LoggingConsumerRebalanceListener
+import org.slf4j.Logger
+
+/**
+ * The [EventSourceCordaConsumerFactory] is responsible for building a configured instance of a CordaConsumer
+ */
+@Suppress("LongParameterList")
+class EventSourceCordaConsumerFactory<K : Any, V : Any>(
+    private val group: String,
```
Comment (yift-r3) [inline libs/messaging/messaging-impl/src/main/kotlin/net/corda/messaging/subscription/factory/EventSourceCordaConsumerFactory.kt:18]: We have the triplet group, clientId and topic together as arguments a few times together. Should we combine them into a data class or just pass the config?

File: libs/messaging/messaging-impl/src/main/kotlin/net/corda/messaging/subscription/factory/EventSourceCordaConsumerFactory.kt [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+        } else {
+            ForwardingRebalanceListener(topic, clientId, partitionAssignmentListener)
+        }
+
+        return cordaConsumerBuilder.createConsumer<K, V>(
+            consumerConfig,
+            messageBusConfig,
+            keyClass,
+            valueClass,
+            { _ ->
```
Comment (yift-r3) [inline libs/messaging/messaging-impl/src/main/kotlin/net/corda/messaging/subscription/factory/EventSourceCordaConsumerFactory.kt]: Maybe `onSerializationError = ...` to make it easier to read?

File: libs/messaging/messaging/src/main/kotlin/net/corda/messaging/api/processor/EventLogProcessor.kt [hunk 10]
```diff
@@ -4,11 +4,11 @@ import net.corda.messaging.api.records.EventLogRecord
 import net.corda.messaging.api.records.Record
 
 /**
- * This interface defines a processor of events from a [EventLogSubscription] on a feed with keys of type [K],
+ * This interface defines a processor of events from a [EventSourceSubscription] on a feed with keys of type [K],
```
Comment (owenstanford) [inline libs/messaging/messaging/src/main/kotlin/net/corda/messaging/api/processor/EventLogProcessor.kt]: Needs to be reverted

File: libs/messaging/messaging-impl/src/test/kotlin/net/corda/messaging/stubs/StubEventSourceProcessor.kt [hunk 11]
```diff
@@ -0,0 +1,21 @@
+package net.corda.messaging.stubs
+
+import net.corda.messaging.api.processor.EventSourceProcessor
+import net.corda.messaging.api.records.EventLogRecord
+import java.util.concurrent.CountDownLatch
+
+class StubEventSourceProcessor<K: Any, V: Any>(private val invocationLatch: CountDownLatch,
```
Comment (owenstanford) [inline libs/messaging/messaging-impl/src/test/kotlin/net/corda/messaging/stubs/StubEventSourceProcessor.kt]: Not sure this class is required, double check

File: libs/messaging/messaging-impl/src/test/kotlin/net/corda/messaging/subscription/EventLogSubscriptionImplTest.kt
Comment (yift-r3) [inline libs/messaging/messaging-impl/src/test/kotlin/net/corda/messaging/subscription/EventLogSubscriptionImplTest.kt:1]: Why was this file deleted?

File: libs/messaging/messaging/src/main/kotlin/net/corda/messaging/api/processor/EventLogProcessor.kt (hunk 10 above)
Comment (yift-r3) [inline libs/messaging/messaging/src/main/kotlin/net/corda/messaging/api/processor/EventLogProcessor.kt]: ?

**Review by yift-r3 — COMMENTED** [review]
Mostly nitpicking comments...

File: libs/messaging/kafka-message-bus-impl/src/main/resources/kafka-messaging-defaults.conf [hunk 12]
```diff
@@ -96,6 +96,12 @@ roles {
             batch.size = 750000
         }
     }
+    eventSource{
+        consumer = ${consumer}{
+            # By default the event source should handle it's own offset commits
```
Comment (driessamyn) [inline libs/messaging/kafka-message-bus-impl/src/main/resources/kafka-messaging-defaults.conf]:
```suggestion
            # By default the event source should handle its own offset commits
```

File: libs/messaging/messaging-impl/src/main/kotlin/net/corda/messaging/subscription/EventSourceConsumer.kt (hunk 1 above)
Comment (owenstanford) [reply libs/messaging/messaging-impl/src/main/kotlin/net/corda/messaging/subscription/EventSourceConsumer.kt]: agree changed

File: libs/messaging/messaging-impl/src/main/kotlin/net/corda/messaging/subscription/factory/CordaSubscriptionFactory.kt (hunk 7 above)
Comment (owenstanford) [reply libs/messaging/messaging-impl/src/main/kotlin/net/corda/messaging/subscription/factory/CordaSubscriptionFactory.kt:197]: unit testing, for these classes there are test cases were we want to assert that some classes of exceptions are logged.

File: libs/messaging/messaging-impl/src/main/kotlin/net/corda/messaging/subscription/factory/EventSourceCordaConsumerFactory.kt (hunk 8 above)
Comment (owenstanford) [reply libs/messaging/messaging-impl/src/main/kotlin/net/corda/messaging/subscription/factory/EventSourceCordaConsumerFactory.kt:18]: I think the individual args are cleaner than the existing overloaded config class, didn't feel worth creating a data class given the limited usage.

File: libs/messaging/messaging-impl/src/test/kotlin/net/corda/messaging/subscription/EventLogSubscriptionImplTest.kt
Comment (owenstanford) [reply libs/messaging/messaging-impl/src/test/kotlin/net/corda/messaging/subscription/EventLogSubscriptionImplTest.kt:1]: mistake

File: libs/messaging/messaging-impl/src/main/kotlin/net/corda/messaging/subscription/consumer/listener/OffsetProviderListener.kt [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+        partitionAssignmentListener?.onPartitionsUnassigned(partitions.toTopicPartitionList())
+    }
+
+    override fun onPartitionsAssigned(partitions: Collection<CordaTopicPartition>) {
+        logAction("assigned", partitions)
+
+        partitionAssignmentListener?.onPartitionsAssigned(partitions.toTopicPartitionList())
+
+        if (offsetProvider != null) {
+            var inputTopicPartitions = partitions.toTopicPartitionList().toSet()
```
Comment (yift-r3) [inline libs/messaging/messaging-impl/src/main/kotlin/net/corda/messaging/subscription/consumer/listener/OffsetProviderListener.kt:35]: `var` -> `val` X2

File: libs/messaging/messaging-impl/src/main/kotlin/net/corda/messaging/subscription/consumer/listener/OffsetProviderListener.kt [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+            var topicPartitionOffsets = offsetProvider.getStartingOffsets(inputTopicPartitions)
```
Comment (yift-r3) [inline libs/messaging/messaging-impl/src/main/kotlin/net/corda/messaging/subscription/consumer/listener/OffsetProviderListener.kt:36]: If we change it to ask for one pair only, we can avoid the null check

File: libs/messaging/messaging-impl/src/main/kotlin/net/corda/messaging/subscription/consumer/listener/OffsetProviderListener.kt [hunk 15]
```diff
@@ -0,0 +1,59 @@
...
+    }
+
+    private fun logAction(action: String, partitions: Collection<CordaTopicPartition>) {
+        partitions.groupBy { it.topic }.forEach { topicGrp ->
+            val partitionIds = topicGrp.value.map { it.partition }.joinToString(",")
+            log.info("Partitions $action for topic ${topicGrp.key}: $partitionIds.")
+        }
+    }
+
+    private fun Collection<CordaTopicPartition>.toTopicPartitionList(): List<Pair<String, Int>> {
```
Comment (yift-r3) [inline libs/messaging/messaging-impl/src/main/kotlin/net/corda/messaging/subscription/consumer/listener/OffsetProviderListener.kt:56]: We use this pair to indicate topics and partitions in many places. Wouldn't it be better to have a data class? (and, TBH, do we need the topic? Isn't it always just a list of partitions of the same topic?)

Not for this PR, of course...

File: libs/messaging/messaging-impl/src/main/kotlin/net/corda/messaging/subscription/factory/CordaSubscriptionFactory.kt [hunk 16]
```diff
@@ -30,18 +32,24 @@ import net.corda.messaging.constants.SubscriptionType
...
+import net.corda.messaging.utils.toEventLogRecord
 import net.corda.schema.configuration.BootConfig.INSTANCE_ID
 import net.corda.schema.configuration.MessagingConfig.MAX_ALLOWED_MSG_SIZE
 import net.corda.web.api.WebServer
 import org.osgi.service.component.annotations.Activate
 import org.osgi.service.component.annotations.Component
 import org.osgi.service.component.annotations.Reference
-import java.util.UUID
+import org.slf4j.LoggerFactory
+import java.util.*
```
Comment (yift-r3) [inline libs/messaging/messaging-impl/src/main/kotlin/net/corda/messaging/subscription/factory/CordaSubscriptionFactory.kt:52]: I think we are only using the UUID

File: libs/messaging/messaging-impl/src/main/kotlin/net/corda/messaging/subscription/factory/EventSourceCordaConsumerFactory.kt (hunk 8 above)
Comment (yift-r3) [reply libs/messaging/messaging-impl/src/main/kotlin/net/corda/messaging/subscription/factory/EventSourceCordaConsumerFactory.kt:18]: The number of arguments, especially since many of them are strings and the order is not apparent (to me)... I will find it confusing if I have to use this constructor in a few months. I know it's a nitpick, but I would try to reduce the number of arguments (on the other hand, we have this pattern all over the place, and making the group a class and the topic an enum would make it event better).

File: libs/messaging/messaging/src/main/kotlin/net/corda/messaging/api/subscription/factory/SubscriptionFactory.kt [hunk 17]
```diff
@@ -130,6 +132,23 @@ interface SubscriptionFactory {
...
+     * events
+     * @param partitionAssignmentListener a listener that reacts to partition assignment and revocations.
+     * @param consumerOffsetProvider a provider of starting offsets for topics and partitions.
+     */
+    fun <K : Any, V : Any> createEventSourceSubscription(
+        subscriptionConfig: SubscriptionConfig,
+        processor: EventSourceProcessor<K, V>,
+        messagingConfig: SmartConfig,
+        partitionAssignmentListener: PartitionAssignmentListener?,
+        consumerOffsetProvider: ConsumerOffsetProvider?
```
Comment (yift-r3) [inline libs/messaging/messaging/src/main/kotlin/net/corda/messaging/api/subscription/factory/SubscriptionFactory.kt:149]: I'm still determining the name, but I don't have a better one. Maybe `OffsetsStartSupplier`?

File: libs/messaging/messaging/src/main/kotlin/net/corda/messaging/api/subscription/listener/ConsumerOffsetProvider.kt [hunk 18]
```diff
@@ -0,0 +1,17 @@
...
+ * in scenarios where the client code want's to manage the consumers itself.
+*/
+interface ConsumerOffsetProvider {
+
+    /**
+     * Called by the consumer whenever it needs to know the starting offsets for a set of topic partitions.
+     *
+     * Note: This method will always be called on the polling thread.
+     * @param topicPartitions A list of topic partitions to provide offsets for
+     * @return A list of offsets for each requested topic partition.
```
Comment (yift-r3) [inline libs/messaging/messaging/src/main/kotlin/net/corda/messaging/api/subscription/listener/ConsumerOffsetProvider.kt:14]: We should state that the return value must contain all the partitions in the request.

File: libs/messaging/messaging/src/main/kotlin/net/corda/messaging/api/subscription/listener/ConsumerOffsetProvider.kt [hunk 19]
```diff
@@ -0,0 +1,17 @@
...
+ * The [ConsumerOffsetProvider] allows a consumer client to provide a custom source of starting offsets. This is useful
+ * in scenarios where the client code want's to manage the consumers itself.
+*/
+interface ConsumerOffsetProvider {
+
+    /**
+     * Called by the consumer whenever it needs to know the starting offsets for a set of topic partitions.
+     *
+     * Note: This method will always be called on the polling thread.
+     * @param topicPartitions A list of topic partitions to provide offsets for
```
Comment (yift-r3) [inline libs/messaging/messaging/src/main/kotlin/net/corda/messaging/api/subscription/listener/ConsumerOffsetProvider.kt:13]: The code has a set, the comment list, maybe both can be a collection?

File: testing/p2p/inmemory-messaging-impl/src/main/kotlin/net/corda/messaging/emulation/subscription/factory/InMemSubscriptionFactory.kt [hunk 20]
```diff
@@ -128,6 +131,23 @@ internal class InMemSubscriptionFactory @Activate constructor(
...
     }
 
+    override fun <K : Any, V : Any> createEventSourceSubscription(
+        subscriptionConfig: SubscriptionConfig,
+        processor: EventSourceProcessor<K, V>,
+        messagingConfig: SmartConfig,
+        partitionAssignmentListener: PartitionAssignmentListener?,
+        consumerOffsetProvider: ConsumerOffsetProvider?
+    ): Subscription<K, V> {
+        return EventSourceSubscription(
```
Comment (yift-r3) [inline testing/p2p/inmemory-messaging-impl/src/main/kotlin/net/corda/messaging/emulation/subscription/factory/InMemSubscriptionFactory.kt:141]: We are not using the `consumerOffsetProvider`. Should we add a comment to remind ourselves it was a conscious decision?

**Review by yift-r3 — APPROVED** [review]
